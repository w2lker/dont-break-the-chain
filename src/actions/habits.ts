
import { List } from 'immutable';
import { IHabit, IHabitsError } from '../models/habit';

export enum HabitsGetStatuses {
  active= 'HABITS/GET/ACTIVE',
  error = 'HABITS/GET/ERROR',
  success = 'HABITS/GET/SUCCESS',
}

interface IGetHabitsActive {
  type: typeof HabitsGetStatuses.active;
}

interface IGetHabitsError {
  type: typeof HabitsGetStatuses.error;
  payload: IHabitsError;
}

interface IGetHabitsSuccess {
  type: typeof HabitsGetStatuses.success;
  // TODO: define data model for response
  payload: List<IHabit>;
}

export function getHabitsStarted() {
  return {
    payload: HabitsGetStatuses.active,
  };
}

export function getHabitsError(errorData: IHabitsError) {
  return {
    type: HabitsGetStatuses.error,
    payload: errorData,
  };
}

export function getHabitsSuccess(habits: any) {
  return {
    type: HabitsGetStatuses.success,
    payload: habits,
  };
}

export type ActionsHabits = IGetHabitsActive | IGetHabitsError | IGetHabitsSuccess;
