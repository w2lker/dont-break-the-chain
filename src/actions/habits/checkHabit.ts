import { ThunkDispatch } from 'redux-thunk';
import { IHabitsError } from '../../models/habit';

export enum HabitCheckStatuses {
  active = 'HABITS/CHECK/ACTIVE',
  error = 'HABITS/CHECK/ERROR',
  success = 'HABITS/CHECK/SUCCESS',
}

interface IHabitCheckPayload {
  id: number;
  checked: boolean;
}

interface IHabitCheckActive {
  type: typeof HabitCheckStatuses.active;
}

interface IHabitCheckError {
  type: typeof HabitCheckStatuses.error;
  payload: IHabitsError;
}

interface IHabitCheckSuccess {
  type: typeof HabitCheckStatuses.success;
  payload: IHabitCheckPayload;
}

export function checkHabitStarted() {
  return {
    type: HabitCheckStatuses.active,
  };
}

export function checkHabitError(errorData: IHabitsError) {
  return {
    type: HabitCheckStatuses.error,
  };
}

export function checkHabitSuccess(id: number, checked: boolean) {
  return {
    type: HabitCheckStatuses.success,
    payload: {
      id,
      checked,
    },
  };
}

export function checkHabit(id: number, desiredStatus: boolean) {
  return (dispatch: ThunkDispatch<any, any, any>) => {
    dispatch(checkHabitStarted());
    // TODO: finalize checking functionality
  };
}
