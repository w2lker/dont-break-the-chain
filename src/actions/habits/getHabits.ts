import { List } from 'immutable';
import { ThunkDispatch } from 'redux-thunk';

import fakeApi from '../../api/fakeApi';
import { IHabit, IHabitsError } from '../../models/habit';

export enum HabitsGetStatuses {
  active = 'HABITS/GET/ACTIVE',
  error = 'HABITS/GET/ERROR',
  success = 'HABITS/GET/SUCCESS',
}

interface IGetHabitsActive {
  type: typeof HabitsGetStatuses.active;
}

interface IGetHabitsError {
  type: typeof HabitsGetStatuses.error;
  payload: IHabitsError;
}

interface IGetHabitsSuccess {
  type: typeof HabitsGetStatuses.success;
  payload: List<IHabit>;
}

export function getHabitsStarted() {
  return {
    type: HabitsGetStatuses.active,
  };
}

export function getHabitsError(errorData: IHabitsError) {
  return {
    type: HabitsGetStatuses.error,
    payload: errorData,
  };
}

export function getHabitsSuccess(habits: any) {
  return {
    type: HabitsGetStatuses.success,
    payload: habits,
  };
}

export function getHabits() {
  return (dispatch: ThunkDispatch<any, any, any>) => {
    dispatch(getHabitsStarted());
    return fakeApi.getHabits().then((response) => {
      // @ts-ignore
      dispatch(getHabitsSuccess(response.data));
    }).catch((response) => {
      dispatch(getHabitsError(response.error));
    });
  };
}

export type GetHabitsActions = IGetHabitsActive | IGetHabitsError | IGetHabitsSuccess;
