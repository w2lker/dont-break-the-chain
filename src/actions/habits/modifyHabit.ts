import { ThunkDispatch } from 'redux-thunk';
import { IHabit, IHabitsError } from '../../models/habit';

export enum HabitModificationStatuses {
  active = 'HABITS/CHECK/ACTIVE',
  error = 'HABITS/CHECK/ERROR',
  success = 'HABITS/CHECK/SUCCESS',
}

interface IHabitModificationActive {
  type: typeof HabitModificationStatuses.active;
}

interface IHabitModificationError {
  type: typeof HabitModificationStatuses.error;
  payload: IHabitsError;
}

interface IHabitModificationSuccess {
  type: typeof HabitModificationStatuses.success;
  payload: IHabit;
}

export function habitModificationStarted(): IHabitModificationActive {
  return {
    type: HabitModificationStatuses.active,
  };
}

export function habitModificationError(errorData: IHabitsError): IHabitModificationError {
  return {
    type: HabitModificationStatuses.error,
    payload: errorData,
  };
}

export function habitModificationSuccess(payload: IHabit): IHabitModificationSuccess {
  return {
    payload,
    type: HabitModificationStatuses.success,
  };
}

export function habitCompleteSwitch(id: number, desiredStatus: boolean) {
  return (dispatch: ThunkDispatch<any, any, any>) => {
    dispatch(habitModificationStarted());
    // TODO: finalize checking functionality
  };
}
