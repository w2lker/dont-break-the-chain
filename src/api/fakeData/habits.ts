import { List } from 'immutable';
import moment from 'moment';

import { habitStatus, IHabit, IHabitDate } from '../../models/habit';

function filledDates(endDate: string, amount: number): List<IHabitDate> {
  const firstDay = List([{
    date: endDate,
    status: habitStatus.end,
    score: amount,
  }]);

  const middleDays = Array(amount - 1)
    .map((item, index) => {
      return {
        date: moment(endDate).subtract(index + 1, 'd').toISOString(),
        status: habitStatus.complete,
        score: amount - index + 1,
      };
    });

  return firstDay.concat(middleDays).push({
    date: moment(endDate).subtract(amount - 1, 'd').toISOString(),
    status: habitStatus.start,
    score: 1,
  });
}

function singleEmptyDate(date: string, today= false): IHabitDate {
  return {
    date,
    status: today ? habitStatus.incompleteToday : habitStatus.incomplete,
    score: 0,
  };
}

function emptyDates(endDate: string, amount: number): List<IHabitDate> {

  const emptyArray: IHabitDate[] = Array(amount - 1)
    .map((item, index) => {
      return singleEmptyDate(moment(endDate).subtract(index, 'd').toISOString());
    });

  return List(emptyArray);
}

const habit1: IHabit = {
  id: 1,
  name: 'Morning runs',
  icon: 'run-fast',
  color: 'blue',
  currentChain: 12,
  longestChain: 45,
  dates: List().concat(
    filledDates(moment().toISOString(), 12),
    emptyDates(moment().subtract(12, 'd').toISOString(), 2),
    filledDates(moment().subtract(14, 'd').toISOString(), 11),
    emptyDates(moment().subtract(25, 'd').toISOString(), 3),
    filledDates(moment().subtract(28, 'd').toISOString(), 45),
  ),
};

const habit2: IHabit = {
  id: 2,
  name: '1h reading',
  icon: 'newspaper',
  color: 'orange',
  currentChain: 35,
  longestChain: 35,
  dates: List([singleEmptyDate(moment().toISOString(), true)]).concat(
    filledDates(moment().subtract(1, 'd').toISOString(), 35),
    emptyDates(moment().subtract(35, 'd').toISOString(), 3),
    filledDates(moment().subtract(38, 'd').toISOString(), 9),
  ),
};
const habit3: IHabit = {
  id: 15,
  name: 'Spend time with family',
  icon: 'account-group',
  color: 'pink',
  currentChain: 8,
  longestChain: 8,
  dates: List([singleEmptyDate(moment().toISOString(), true)]).concat(
    filledDates(moment().subtract(1, 'd').toISOString(), 8),
  ),
};

const habits: List<IHabit> = List([habit1, habit2, habit3]);

export default habits;
