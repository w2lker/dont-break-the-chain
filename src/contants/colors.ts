import colorBlue from '@material-ui/core/colors/blue';
import colorBrown from '@material-ui/core/colors/brown';
import colorCyan from '@material-ui/core/colors/cyan';
import colorDeepPurple from '@material-ui/core/colors/deepPurple';
import colorGreen from '@material-ui/core/colors/green';
import colorIndigo from '@material-ui/core/colors/indigo';
import colorOrange from '@material-ui/core/colors/orange';
import colorPink from '@material-ui/core/colors/pink';
import colorPurple from '@material-ui/core/colors/purple';
import colorTeal from '@material-ui/core/colors/teal';

export type colorsKeys = 'pink' | 'purple' | 'deepPurple' | 'indigo' | 'blue' | 'cyan' | 'teal' | 'green' | 'orange' | 'brown';

export const colors: {[color: string]: any}  = {
  blue: colorBlue,
  brown: colorBrown,
  pink: colorPink,
  cyan: colorCyan,
  deepPurple: colorDeepPurple,
  indigo: colorIndigo,
  green: colorGreen,
  orange: colorOrange,
  purple: colorPurple,
  teal: colorTeal,
};
