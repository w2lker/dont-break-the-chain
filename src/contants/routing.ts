export const routing = {
  default: '/',
  chains: '/chains',
  habits: '/habit',
  profile: '/account',
};
