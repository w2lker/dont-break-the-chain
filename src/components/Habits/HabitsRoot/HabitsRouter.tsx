import React from 'react';

import { Route, RouteComponentProps } from 'react-router-dom';

import HabitsPage from '../HabitsPage';
import HabitCreate from '../HabitsPage/HabitCreate';

import { routing } from '../../../contants/routing';

export interface IHabitsRootProps extends RouteComponentProps<any> {}

const HabitsRouter: React.FC<IHabitsRootProps> = (props) => {
  const {} = props;
  const habitsRouting = routing.habits;
  return (
    <React.Fragment>
      <Route component={HabitsPage} />
      <Route path={habitsRouting.root + habitsRouting.create} component={HabitCreate} />
      <Route path={habitsRouting.root + habitsRouting.update} component={HabitCreate} />
    </React.Fragment>
  );
};

export default HabitsRouter;
