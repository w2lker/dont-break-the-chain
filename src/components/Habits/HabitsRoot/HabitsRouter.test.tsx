import React from 'react';

import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import { MemoryRouter, Route } from 'react-router-dom';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';

import HabitsRouter from './HabitsRouter';

import habits from '../../../api/fakeData/habits';
import { routing } from '../../../contants/routing';

const testSetup = () => {
  const middleware = [thunk];
  const mockStore = configureMockStore(middleware);
  // TODO: add sample data for habits page
  const sampleStoreData = {
    habits: {
      habits,
    },
  };
  const store = mockStore(sampleStoreData);
  const wrappedComponent = (url: string) => (
    <Provider store={store}>
      <MemoryRouter initialEntries={[url]}>
        <Route component={HabitsRouter} />
      </MemoryRouter>
    </Provider>
  );
  const rootUrl = routing.habits.root;
  const createUrl = routing.habits.root + routing.habits.create;
  return { wrappedComponent, rootUrl, createUrl };
};

it('Match snapshot', () => {
  const { wrappedComponent, rootUrl, createUrl } = testSetup();
  // Root habit page
  const rootComponent = mount(wrappedComponent(rootUrl));
  expect(rootComponent.debug()).toMatchSnapshot();
  // Create habit page
  const createHabitComponent = mount(wrappedComponent(createUrl));
  expect(createHabitComponent.debug()).toMatchSnapshot();
});

it('Renders Habits root page', () => {
  const { wrappedComponent, rootUrl } = testSetup();
  const component = mount(wrappedComponent(rootUrl));
  const page = component.find('HabitsPage');
  expect(page.length).toBe(1);
});

it('Renders Create Habit page', () => {
  const { wrappedComponent, createUrl } = testSetup();
  const component = mount(wrappedComponent(createUrl));
  const createPage = component.find('HabitCreate');
  expect(createPage.length).toBe(1);
});
