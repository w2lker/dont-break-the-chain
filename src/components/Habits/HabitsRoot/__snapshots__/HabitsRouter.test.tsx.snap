// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Match snapshot 1`] = `
"<Provider store={{...}}>
  <MemoryRouter initialEntries={{...}}>
    <Router history={{...}}>
      <Route component={[Function: HabitsRouter]}>
        <HabitsRouter history={{...}} location={{...}} match={{...}} staticContext={[undefined]}>
          <Route component={{...}}>
            <WithStyles(HabitsPage) history={{...}} location={{...}} match={{...}} staticContext={[undefined]}>
              <HabitsPage classes={{...}} history={{...}} location={{...}} match={{...}} staticContext={[undefined]}>
                <div className=\\"HabitsPage-root-1\\">
                  <WithStyles(HabitsHeader)>
                    <HabitsHeader classes={{...}}>
                      <div className=\\"HabitsHeader-root-2\\">
                        <h4 className=\\"HabitsHeader-title-3\\">
                          My Habits
                        </h4>
                      </div>
                    </HabitsHeader>
                  </WithStyles(HabitsHeader)>
                </div>
              </HabitsPage>
            </WithStyles(HabitsPage)>
          </Route>
          <Route path=\\"/habit/create\\" component={{...}} />
          <Route path=\\"/habit/update/id\\" component={{...}} />
        </HabitsRouter>
      </Route>
    </Router>
  </MemoryRouter>
</Provider>"
`;

exports[`Match snapshot 2`] = `
"<Provider store={{...}}>
  <MemoryRouter initialEntries={{...}}>
    <Router history={{...}}>
      <Route component={[Function: HabitsRouter]}>
        <HabitsRouter history={{...}} location={{...}} match={{...}} staticContext={[undefined]}>
          <Route component={{...}}>
            <WithStyles(HabitsPage) history={{...}} location={{...}} match={{...}} staticContext={[undefined]}>
              <HabitsPage classes={{...}} history={{...}} location={{...}} match={{...}} staticContext={[undefined]}>
                <div className=\\"HabitsPage-root-1\\">
                  <WithStyles(HabitsHeader)>
                    <HabitsHeader classes={{...}}>
                      <div className=\\"HabitsHeader-root-2\\">
                        <h4 className=\\"HabitsHeader-title-3\\">
                          My Habits
                        </h4>
                      </div>
                    </HabitsHeader>
                  </WithStyles(HabitsHeader)>
                </div>
              </HabitsPage>
            </WithStyles(HabitsPage)>
          </Route>
          <Route path=\\"/habit/create\\" component={{...}}>
            <WithStyles(HabitCreate) history={{...}} location={{...}} match={{...}} staticContext={[undefined]}>
              <HabitCreate classes={{...}} history={{...}} location={{...}} match={{...}} staticContext={[undefined]}>
                <div className=\\"HabitCreate-root-4\\">
                  HabitCreate Component
                </div>
              </HabitCreate>
            </WithStyles(HabitCreate)>
          </Route>
          <Route path=\\"/habit/update/id\\" component={{...}} />
        </HabitsRouter>
      </Route>
    </Router>
  </MemoryRouter>
</Provider>"
`;
